File: Services/WallhavenService.cs
using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;
using AnimeWallpaper.Models;

namespace AnimeWallpaper.Services;

public interface IWallhavenService
{
    Task<List<WallhavenItem>> SearchWallpapersAsync(AppConfig config, CancellationToken cancellationToken = default);
}

public class WallhavenService : IWallhavenService
{
    private readonly HttpClient _httpClient;
    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNameCaseInsensitive = true
    };

    public WallhavenService(HttpClient? httpClient = null)
    {
        _httpClient = httpClient ?? new HttpClient();
        _httpClient.Timeout = TimeSpan.FromSeconds(30);
        if (_httpClient.DefaultRequestHeaders.UserAgent.Count == 0)
        {
            _httpClient.DefaultRequestHeaders.UserAgent.Add(
                new ProductInfoHeaderValue("AnimeWallpaperApp", "1.0")
            );
        }
    }

    public async Task<List<WallhavenItem>> SearchWallpapersAsync(AppConfig config, CancellationToken cancellationToken = default)
    {
        try
        {
            var query = string.IsNullOrWhiteSpace(config.Query) ? "anime" : config.Query.Trim();
            var categories = "010"; // 010 = Anime category
            var purity = config.SfwOnly ? "100" : "110"; // 100 = SFW only, 110 = SFW + Sketchy
            var sorting = config.Random ? "random" : "toplist";
            var minWidth = config.MinimumWidth > 0 ? config.MinimumWidth : 1920;
            var minHeight = config.MinimumHeight > 0 ? config.MinimumHeight : 1080;
            var atleast = $"{minWidth}x{minHeight}";

            var uriBuilder = new UriBuilder("https://wallhaven.cc/api/v1/search");
            var queryParams = new List<string>
            {
                $"q={Uri.EscapeDataString(query)}",
                $"categories={categories}",
                $"purity={purity}",
                $"sorting={sorting}",
                $"atleast={atleast}"
            };

            if (!string.IsNullOrWhiteSpace(config.WallhavenApiKey))
            {
                queryParams.Add($"apikey={Uri.EscapeDataString(config.WallhavenApiKey.Trim())}");
            }

            uriBuilder.Query = string.Join("&", queryParams);

            Logger.Info($"Requesting Wallhaven: q='{query}', categories={categories}, purity={purity}, sorting={sorting}, atleast={atleast}");

            using var request = new HttpRequestMessage(HttpMethod.Get, uriBuilder.Uri);

            // Also attach header if API key is provided
            if (!string.IsNullOrWhiteSpace(config.WallhavenApiKey))
            {
                request.Headers.Add("X-API-Key", config.WallhavenApiKey.Trim());
            }

            using var response = await _httpClient.SendAsync(request, cancellationToken);

            if ((int)response.StatusCode == 429)
            {
                Logger.Warn("Wallhaven API rate limit hit (HTTP 429). Will retry on next cycle.");
                return new List<WallhavenItem>();
            }

            if (!response.IsSuccessStatusCode)
            {
                Logger.Warn($"Wallhaven API returned non-success status: {(int)response.StatusCode} {response.ReasonPhrase}");
                return new List<WallhavenItem>();
            }

            var content = await response.Content.ReadAsStringAsync(cancellationToken);
            var result = JsonSerializer.Deserialize<WallhavenResponse>(content, JsonOptions);

            if (result?.Data == null || result.Data.Count == 0)
            {
                Logger.Warn("Wallhaven API returned 0 results for the current query.");
                return new List<WallhavenItem>();
            }

            Logger.Info($"Wallhaven returned {result.Data.Count} wallpapers.");
            return result.Data;
        }
        catch (OperationCanceledException)
        {
            Logger.Info("Wallhaven search request cancelled.");
            return new List<WallhavenItem>();
        }
        catch (HttpRequestException ex)
        {
            Logger.Error("Network error during Wallhaven request. Check internet connection.", ex);
            return new List<WallhavenItem>();
        }
        catch (Exception ex)
        {
            Logger.Error("Unexpected error querying Wallhaven API.", ex);
            return new List<WallhavenItem>();
        }
    }
}

File: macos/Sources/Services/WallhavenService.swift
import Foundation

public protocol WallhavenServiceProtocol: Sendable {
    func searchWallpapers(config: AppConfig) async throws -> [WallhavenItem]
}

public final class WallhavenService: WallhavenServiceProtocol, @unchecked Sendable {
    private let session: URLSession

    public init(session: URLSession = .shared) {
        self.session = session
    }

    public func searchWallpapers(config: AppConfig) async throws -> [WallhavenItem] {
        let query = config.query.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ? "anime" : config.query.trimmingCharacters(in: .whitespacesAndNewlines)
        let categories = "010" // Anime only
        let purity = config.sfwOnly ? "100" : "110" // 100 = SFW
        let sorting = config.random ? "random" : "toplist"
        let minW = config.minimumWidth > 0 ? config.minimumWidth : 1920
        let minH = config.minimumHeight > 0 ? config.minimumHeight : 1080
        let atleast = "\(minW)x\(minH)"

        var components = URLComponents(string: "https://wallhaven.cc/api/v1/search")!
        var queryItems: [URLQueryItem] = [
            URLQueryItem(name: "q", value: query),
            URLQueryItem(name: "categories", value: categories),
            URLQueryItem(name: "purity", value: purity),
            URLQueryItem(name: "sorting", value: sorting),
            URLQueryItem(name: "atleast", value: atleast)
        ]

        let apiKey = config.wallhavenApiKey.trimmingCharacters(in: .whitespacesAndNewlines)
        if !apiKey.isEmpty {
            queryItems.append(URLQueryItem(name: "apikey", value: apiKey))
        }

        components.queryItems = queryItems

        guard let url = components.url else {
            Logger.error("Failed to construct Wallhaven search URL")
            return []
        }

        Logger.info("Requesting Wallhaven: q='\(query)', categories=\(categories), purity=\(purity), sorting=\(sorting), atleast=\(atleast)")

        var request = URLRequest(url: url, timeoutInterval: 30)
        request.setValue("AnimeWallpaperMac/1.0", forHTTPHeaderField: "User-Agent")
        if !apiKey.isEmpty {
            request.setValue(apiKey, forHTTPHeaderField: "X-API-Key")
        }

        let (data, response) = try await session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            Logger.warn("Wallhaven returned non-HTTP response")
            return []
        }

        if httpResponse.statusCode == 429 {
            Logger.warn("Wallhaven API rate limit hit (HTTP 429). Will retry next cycle.")
            return []
        }

        guard (200...299).contains(httpResponse.statusCode) else {
            Logger.warn("Wallhaven API returned error status: \(httpResponse.statusCode)")
            return []
        }

        let decoded = try JSONDecoder().decode(WallhavenResponse.self, from: data)
        guard let items = decoded.data, !items.isEmpty else {
            Logger.warn("Wallhaven API returned 0 results for query '\(query)'")
            return []
        }

        Logger.info("Wallhaven returned \(items.count) wallpapers.")
        return items
    }
}
